### PR TITLE
Config api for the daemon

### PR DESCRIPTION
**Pull request by anjannath**
~I did these changes on top of #1221 so those needs to go in before this, when reviewing just review this commit https://github.com/code-ready/crc/pull/1246/commits/149b27f7ab684c210a66bef5720d60fe9d23fb56~
 Part of #723

### CODE REVIEW

**Comment by gbraad** [comment]
> - setconfig example
>   {"command":"setconfig", "args":{"bundle":"path/to/bundle", "cpu":"4"}}
> - getconfig example
>   {"command":"getconfig"} witouth the args object it will return all
>   {"command":"getconfig", "args":{"properties":["bundle", "memory"]}}
> - unsetconfig example
>   {"command":"unsetconfig", "args":{"properties":["bundle"]}}

This makes sense

File: pkg/crc/api/types.go [hunk 1]
```diff
@@ -25,6 +26,20 @@ type clusterOpsRequest struct {
...
-	Args    map[string]string `json:"args,omitempty"`
+	Command string `json:"command"`
+	Args    json.RawMessage
+}
+
+// setOrUnsetConfigResult struct is used to return the result of
+// setconfig/unsetconfig command
+type setOrUnsetConfigResult struct {
+	Error   string
+	Success bool
```
Comment (cfergeau) [inline pkg/crc/api/types.go:37]: Imo it's a mistake that some crc APIs return both an error and a success boolean. If `Success` is true, then `Error` must be `""`, if `Success` is false, then `Error` must be  non-empty, so they are mutually exclusive, so I would either send a message with only an `Error` field, or a message with the command result. No need for `Success`, unless maybe if the response would be empty otherwise.

I wonder if set/unset should return the list of config values it set/unset?

Comment (anjannath) [reply pkg/crc/api/types.go:37]: Yes they are mutually exclusive, only the error would be enough to know if succeeded or failed but for `setconfig` and `unsetconfig` there is no command result, so `Success` is used.

> I wonder if set/unset should return the list of config values it set/unset?

I thought about this, in case any of the set/unset on the provided keys fails, we are getting back an error (err msg contains the name of the keys but harder to extract)

**Comment by anjannath** [comment]
>Why the difference between the 2? Why not {"command":"unsetconfig", "args":{"bundle", "cpu"}}?

For `setconfig` we need a key value pair, but for `unsetconfig` and `getconfig` we just need the keys, so sending them inside the _properties_ array.

File: pkg/crc/config/viper_config.go [hunk 2]
```diff
@@ -153,6 +153,19 @@ func AllConfigKeys() []string {
...
+	for _, key := range AllConfigKeys() {
+		v, err := Get(key)
+		if err != nil {
+			allConfigs[key] = nil
+		}
+		allConfigs[key] = v
+	}
+	return allConfigs
+}
+
```
[the remark this answers is not in the document]
Comment (anjannath) [reply pkg/crc/config/viper_config.go:168]: Yeah, we can get the list of keys from this function. if you think i should remove the `AllConfigKeys` then i can do that, i think right now only `cmd/config` uses it.

File: pkg/crc/api/handlers.go [hunk 3]
```diff
@@ -64,18 +66,135 @@ func getPullSecretFileContent() (string, error) {
...
+		if err != nil {
+			multiError.Collect(err)
+		}
+		err = crcConfig.WriteConfig()
+		if err != nil {
+			multiError.Collect(err)
+		}
+	}
+	if len(multiError.Errors) != 0 {
+		setConfigResult.Error = fmt.Sprintf("%v", multiError.ToError())
```
[the remark this answers is not in the document]
Comment (anjannath) [reply pkg/crc/api/handlers.go:110]: This is inline to what you mention in https://github.com/code-ready/crc/pull/1246#discussion_r429869470 so, `setconfig` will return

```json
{
    "properties": ["bundle", "pull-secret-file"],
    "Error": "xxxxxxxxx"
}
```

or when no errors..

```json
{
    "properties": ["bundle", "pull-secret-file", "memory"],
    "Error": ""
}
```

**Comment by anjannath** [comment]
> > > Why the difference between the 2? Why not {"command":"unsetconfig", "args":{"bundle", "cpu"}}?
> > 
> > 
> > For `setconfig` we need a key value pair, but for `unsetconfig` and `getconfig` we just need the keys, so sending them inside the _properties_ array.
> 
> I don't understand what the `properties` array is bringing us? Why not have a `properties` array in the case of `setconfig` too then?

For `setconfig` we have an object, key/value pair so we can directly provide that to args, but for `unset/getconfig` we just the need the properties'/keys' names, so sending them in a array makes sense, and i named it as _properties_. we cannot pass it like you mentioned.. however we can do this..

```json
{"command":"unsetconfig", "args":{"bundle":"", "cpu":""}}
```

but using an array is better..
```json
{"command":"unsetconfig", "args":{"properties":["bundle", "cpu"]}}}
```

We can also have the _properties_ in `setconfig`, but its just one more layer..

```json
{"command":"setconfig", "args": {"properties": {"bundle":"path/to/bundle", "cpu":4}}}
```

**Comment by cfergeau** [comment]
> > > > Why the difference between the 2? Why not {"command":"unsetconfig", "args":{"bundle", "cpu"}}?
> > > 
> > > 
> > > For `setconfig` we need a key value pair, but for `unsetconfig` and `getconfig` we just need the keys, so sending them inside the _properties_ array.
> > 
> > 
> > I don't understand what the `properties` array is bringing us? Why not have a `properties` array in the case of `setconfig` too then?
> 
> For `setconfig` we have an object, key/value pair so we can directly provide that to args, but for `unset/getconfig` we just the need the properties'/keys' names, so sending them in a array makes sense, and i named it as _properties_. we cannot pass it like you mentioned.. however we can do this..
> 
> ```json
> {"command":"unsetconfig", "args":{"bundle":"", "cpu":""}}
> ```
> 
> but using an array is better..
> 
> ```json
> {"command":"unsetconfig", "args":{"properties":["bundle", "cpu"]}}}
> ```
> 
> We can also have the _properties_ in `setconfig`, but its just one more layer..
> 
> ```json
> {"command":"setconfig", "args": {"properties": {"bundle":"path/to/bundle", "cpu":4}}}
> ```

In one case using an array (which is an extra layer) is better, but in the other case, using an additional layer seems to be a problem? :)

I think in both cases this should be wrapped in `properties`. If we ever want to add additional parameters to the `setconfig` command, the way it's currently done means we would have to use a parameter name which does not conflict with the name of a config key. So better to just wrap this in `properties`, and  be consistent and keep our options open for future extensions of the protocol.

File: pkg/crc/config/viper_config.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (anjannath) [reply pkg/crc/config/viper_config.go:160]: Done!

**Comment by anjannath** [comment]
@cfergeau i think all the review commits are addressed, please have a look again, thanks!

File: pkg/crc/api/handlers.go [hunk 4]
```diff
@@ -64,18 +66,139 @@ func getPullSecretFileContent() (string, error) {
...
 }
 
-func webconsoleURLHandler(_ ArgsType) string {
+func webconsoleURLHandler(_ json.RawMessage) string {
 	consoleConfig := machine.ConsoleConfig{Name: constants.DefaultName}
 	r, _ := machine.GetConsoleURL(consoleConfig)
 	return encodeStructToJSON(r)
 }
 
+func setConfigHandler(args json.RawMessage) string {
```
Comment (guillaumerose) [inline pkg/crc/api/handlers.go:81]: A test in api_test.go would be really good!

File: pkg/crc/api/handlers.go [hunk 5]
```diff
@@ -2,24 +2,26 @@ package api
...
 
-func statusHandler(client machine.Client, _ ArgsType) string {
+func statusHandler(client machine.Client, _ json.RawMessage) string {
 	statusConfig := machine.ClusterStatusConfig{Name: constants.DefaultName}
 	clusterStatus, _ := client.Status(statusConfig)
 	return encodeStructToJSON(clusterStatus)
 }
 
-func stopHandler(client machine.Client, _ ArgsType) string {
+func stopHandler(client machine.Client, _ json.RawMessage) string {
```
Comment (praveenkumar) [inline pkg/crc/api/handlers.go:24]: @anjannath should this be in a different commit message where you changed the `ArgsType` to `json.RawMessage` ?

Comment (anjannath) [reply pkg/crc/api/handlers.go:24]: Yeah, makes sense to separate this,will do thanks!

File: cmd/crc/cmd/config/config.go [hunk 6]
```diff
@@ -77,7 +77,11 @@ func less(lhsKey, rhsKey string) bool {
 
 func configurableFields() string {
 	var fields []string
-	keys := cfg.AllConfigKeys()
+	var keys []string
```
[the remark this answers is not in the document]
Comment (cfergeau) [reply cmd/crc/cmd/config/config.go]: And this causes https://github.com/code-ready/crc/issues/1487 ;)
